### DISCUSSION TITLE
Question about complicated overloaded decorator case

### DISCUSSION THREAD

**Discussion by bijij** [microsoft/pyright#3100] [Q&A] [answered]
I have an interesting decorator I wish to annotate, which depending on whether the decorated function is in a class or not, will affect the return type.

The code ends up being something like this:
```py
from __future__ import annotations

from typing import Any, TYPE_CHECKING, Generic, TypeVar, Callable, overload, Optional

if TYPE_CHECKING:
    from typing_extensions import Concatenate, ParamSpec

    P = ParamSpec('P')
else:
    P = TypeVar('P')


T = TypeVar('T')
_FT = TypeVar('_FT', bound='Optional[Foo]')


class Foo:
    ...


# this type is returned by the decorator
class Bar(Generic[_FT, P, T]):
    ...


# in the case where the decorator is used outside a class _FT is None
@overload
def make_bar() -> Callable[[Callable[P, T]], Bar[None, P, T]]:
    ...


# when used in a class the type should match `self`
@overload
def make_bar() -> Callable[[Callable[Concatenate[_FT, P], T]], Bar[_FT, P, T]]:
    ...


# any return type here just to make things easier.
def make_bar() -> Any:
    ...


@make_bar()
def bar(test: int) -> str:
    ...


# pyright: Type of "bar" is "Bar[None, (test: int), str]"
# this is what I'd expect to see
# if the positions of the overloads are reversed we see:
# Type of "bar" is "Bar[_FT@make_bar, (), str]"
# _FT is unbound, undesirable
reveal_type(bar)


class MyFoo(Foo):
    @make_bar()
    def bar(self, test: int) -> str:
        ...


# pyright: Type of "MyFoo.bar" is "Bar[None, (self: MyFoo, test: int), str]"
# None here is clearly wrong, expected MyFoo
# if the positions of the overloads are reversed we see:
# Type of "MyFoo.bar" is "Bar[MyFoo, (test: int), str]"
# The desired behaviour
reveal_type(MyFoo.bar)

```

Is there any way, with python's current static typing features to achieve the desired behaviour in both cases?

This would easily be possible if `TypeVar` had the notion of a `default` a-la rust's generics
say `_FT = TypeVar('_FT', bound='Optional[Foo]', default=None)` the unbound case would default to None

I'm not sure if the concept of defaults for `TypeVar`s has been discussed in any of the py typing circles, if not I might work on trying to get such a feature going, especially if it'd be the only way to have the behaviour I desire in this case.

Thanks.

**Comment by erictraut** [comment]
I understand the motivation, but this approach feels very fragile. It limits the decorator to classes that derive from `Foo`. It also produces the wrong results outside of a class if the function happened to accept a first parameter of type `Foo` or `None`.

Would it be feasible to provide two decorator variants, one for use with instance methods and another for non-methods (or static methods)? Then you could avoid the need for an overload.

Capturing (or not capturing) `self` with a ParamSpec is a problem that [others have run into](https://github.com/python/typing/discussions/946), and I don't see a good solution. Perhaps @pradeep90 (member of the Pyre team, which was responsible for PEP 612) might have additional thoughts here.

There has been some discussion among Python typing circles of providing support for default TypeVar values, but they haven't gone very far, and they didn't contemplate the use case you have in mind here.

**Reply by bijij** [reply]
Unfortunate, running pyright on strict errors suggesting the two overloads are equivalent, is this because pyright doesn't look into the inner `Callable`? Or because `ParamSpec` is a catch-all?

> I understand the motivation, but this approach feels very fragile.

Not going to disagree. However in my specific circumstance it can't really be avoided.

> It limits the decorator to classes that derive from `Foo`.

In the actual case yes the decorator is only supposed to be used outside any class or inside classes which derive a specific class. 

>  It also produces the wrong results outside of a class if the function happened to accept a first parameter of type `Foo` or None.

In the actual code, the first (non-self) argument of the function has to derive a 3rd type, say `Baz`

So it actually accepts `(Foo, Baz, *P.args, **P.kwargs) -> T | (Baz, *P.args, **P.kwargs) -> T` Not sure if having the first argument type forced like so helps me in any way. 

> Would it be feasible to provide two decorator variants, one for use with instance methods and another for non-methods (or static methods)? Then you could avoid the need for an overload.

This is actually being used to annotate a pre-existing library, which has been by use of a Metaclass providing this behaviour, so changing unfortunately would be breaking.


> There has been some discussion among Python typing circles of providing support for default TypeVar values, but they haven't gone very far, and they didn't contemplate the use case you have in mind here.

Interesting, I'll definitely have a look at what's already been discussed on the topic, wasn't sure if my specific scenario had been covered, but it's the first case I've encountered where I've seen it might be a possible solution.

Thanks.

**Comment by bijij** [comment, marked as answer]
I have found for my purposes having the decorator return a class which itself has its `__call__`  method overloaded solves the problem.

Looks something like this, (I added in the `Baz` to better reflect the actual situation)

```py
from __future__ import annotations

from typing import Any, TYPE_CHECKING, Generic, TypeVar, Callable, overload, Optional

if TYPE_CHECKING:
    from typing_extensions import Concatenate, ParamSpec

    P = ParamSpec('P')
else:
    P = TypeVar('P')


T = TypeVar('T')
_FT = TypeVar('_FT', bound='Optional[Foo]')


class Foo:
    ...


# this type is returned by the decorator
class Bar(Generic[_FT, P, T]):
    ...


class Baz:
    ...


# using a class here allows us to have the overloads on the inner function rather than the outer.
class BarDeco:
    @overload
    def __call__(self, func: Callable[Concatenate[Baz, P], T], /) -> Bar[None, P, T]: ...

    @overload
    def __call__(self, func: Callable[Concatenate[_FT, Baz, P], T], /) -> Bar[_FT, P, T]: ...

    def __call__(self, func: Any, /) -> Any:
       ...


def make_bar() -> BarDeco:
    ...


@make_bar()
def bar(baz: Baz, test: int) -> str:
    ...

# pyright: Type of "bar" is "Bar[None, (test: int), str]"
# looking good
reveal_type(bar)


class MyFoo(Foo):
    @make_bar()
    def bar(self, baz: Baz, test: int) -> str:
        ...

# pyright: Type of "MyFoo.bar" is "Bar[MyFoo, (test: int), str]"
# The desired behaviour
reveal_type(MyFoo.bar)
```

Not the prettiest code, but it'll do what I need it to do.
Unsure if this is supposed to work but I'm going to have to rely on it for now.

**Reply by erictraut** [reply]
That's a clever solution. Looks good to me.
